File: Recursion Maze/Backtracking/Q1. Combinations/Q1_Combinations.java
import java.util.*;

class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>(); // List to store all combinations
        backtrack(1, n, k, new ArrayList<>(), result); // Start backtracking from 1
        return result;
    }

    // Helper method for backtracking
    private void backtrack(int start, int n, int k, List<Integer> tempList, List<List<Integer>> result) {
        if (tempList.size() == k) { // If current combination size equals k
            result.add(new ArrayList<>(tempList)); // Add a copy to result
            return;
        }
        for (int i = start; i <= n; i++) { // Loop through numbers from start to n
            tempList.add(i); // Add current number to combination
            backtrack(i + 1, n, k, tempList, result); // Recurse with next number
            tempList.remove(tempList.size() - 1); // Backtrack, remove last number
        }
    }
}

File: Recursion Maze/Backtracking/Q1. Combinations/Q1_Combinations.py
class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        result = []  # List to store all combinations

        # Helper function for backtracking
        def backtrack(start, path):
            if len(path) == k:  # If combination size equals k
                result.append(list(path))  # Add a copy of current combination
                return
            for i in range(start, n + 1):  # Loop through numbers from start to n
                path.append(i)  # Add current number to combination
                backtrack(i + 1, path)  # Recurse with next number
                path.pop()  # Backtrack, remove last number

        backtrack(1, [])  # Start backtracking from 1 with empty path
        return result
